[iOS] Update WKSETextInput/WKSETextInteraction adoption

whsieh opened the pull request:
#### 27dbd349074d620e5080dafbc5d80ee760de25f6
<pre>
[iOS] Update WKSETextInput/WKSETextInteraction adoption
<a href="https://bugs.webkit.org/show_bug.cgi?id=267555">https://bugs.webkit.org/show_bug.cgi?id=267555</a>
<a href="https://rdar.apple.com/121011826">rdar://121011826</a>

Reviewed by Tim Horton.

Since 272018@main landed, there have been numerous minor adjustments to the async text input API in
ServiceExtensions; now that the API shape is finally stable, we adjust the WebKit preadoption to use
these new method and symbol names when `SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` is set, while
ensuring binary compatibility with the old UIKit async text input symbols.

See below for more details.

* Source/WebCore/platform/ios/WebEvent.mm:
(webEventType):
(isChangingKeyModifiers):
(-[WebEvent initWithKeyEntry:]):
(-[WebEvent originalKeyEntry]):
(-[WebEvent initWithKeyEvent:]): Deleted.
(-[WebEvent originalKeyEvent]): Deleted.

Rename `WebSEKeyEvent` to `WebSEKeyEntry`, and `WebSEKeyEventType` to `WebSEKeyPressState`.

* Source/WebCore/platform/ios/WebEventPrivate.h:
* Source/WebCore/platform/ios/WebSEDefinitions.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove `HAVE_UI_ASYNC_TEXT_INTERACTION_DELEGATE`, since it&apos;s been in SDKs where
`HAVE_UI_ASYNC_TEXT_INTERACTION` is set for months now.

* Source/WebKit/Shared/DocumentEditingContext.mm:

Import the relevant system headers.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Rename numerous text input delegate method implementations, to fit the new API shape.

(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):

Also address the fact that calls to `-hasSelectablePositionAtPoint:` have been replaced with calls
to `-textInteractionGesture:shouldBeginAtPoint:` with `WKSEGestureTypeForceTouch`, since the former
was only used in a few places for the purposes of handling force touch.

(-[WKContentView replaceForWebView:]):
(-[WKContentView selectionBarColor]):

Replace `-selectionBarColor` with `-selectionHandleColor`.

(-[WKContentView canPerformAction:withSender:]):

Replace `-moveInDirection:byGranularity:` with `-moveInStorageDirection:byGranularity:`.

(toGestureType):
(toWKSEGestureType):
(toSelectionTouch):
(toWKSESelectionTouchPhase):
(selectionChangedWithTouch):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(logTextInteraction):
(-[WKContentView _internalBeginSelectionChange]):
(-[WKContentView _internalEndSelectionChange]):
(-[WKContentView addTextAlternatives:]):
(-[WKContentView _updateTextInputTraits:]):
(shiftKeyState):
(-[WKContentView modifierFlagsDidChangeFrom:to:]):
(-[WKContentView _deferKeyEventToInputMethodEditing:]):
(-[WKContentView _interpretKeyEvent:isCharEvent:]):
(-[WKContentView _transpose]):
(-[WKContentView _updateInputContextAfterBlurringAndRefocusingElement]):
(-[WKContentView _selectionChanged]):
(-[WKContentView _provideSuggestionsToInputDelegate:]):
(-[WKContentView _provideUITextSuggestionsToInputDelegate:]):
(toWebDocumentRequestOptions):
(toWebRequest):
(-[WKContentView requestDocumentContext:completionHandler:]):
(-[WKContentView asWKSETextInput]):

The following method implementations represent the set of final API names, which wrap calls to
`UIAsyncTextInput` methods.

(-[WKContentView systemWillPresentEditMenuWithAnimator:]):
(-[WKContentView systemWillDismissEditMenuWithAnimator:]):
(-[WKContentView insertTextAlternatives:]):
(-[WKContentView insertText:textAlternatives:style:]):

Note: `-insertText:textAlternatives:style:` was renamed to `-insertTextAlternatives:` here, but we
implement both versions of this new delegate method for now to ensure binary compatibility with
both versions of ServiceExtensions.

(-[WKContentView isPointNearMarkedText:]):
(-[WKContentView replaceSelectedText:withText:]):
(-[WKContentView updateCurrentSelectionTo:fromGesture:inState:]):
(-[WKContentView setSelectionFromPoint:toPoint:gesture:state:]):
(-[WKContentView adjustSelectionBoundaryToPoint:touchPhase:baseIsStart:flags:]):
(-[WKContentView moveSelectionAtBoundary:inStorageDirection:completionHandler:]):
(-[WKContentView extendedTextInputTraits]):
(-[WKContentView selectTextForEditMenuWithLocationInView:completionHandler:]):
(-[WKContentView isSelectionAtDocumentStart]):
(-[WKContentView automaticallyPresentEditMenu]):
(-[WKContentView asyncSystemInputDelegate]):
(-[WKContentView setAsyncInputDelegate:setAsyncSystemInputDelegate:]):
(-[WKContentView handleKeyEntry:withCompletionHandler:handleAsyncKeyEvent:withCompletionHandler:]):
(-[WKContentView replaceText:withText:options:completionHandler:replaceText:withText:options:withCompletionHandler:]):
(-[WKContentView moveInStorageDirection:byGranularity:moveInDirection:byGranularity:]):
(-[WKContentView extendInStorageDirection:byGranularity:extendInDirection:byGranularity:]):
(-[WKContentView adjustSelectionByRange:completionHandler:adjustSelection:completionHandler:]):
(-[WKContentView transposeCharacters]):
(-[WKContentView shiftKeyStateChangedFromState:toState:shiftKeyStateChangedFrom:to:]):
(-[WKContentView systemWillChangeSelectionForInteraction:selectionWillChange:]):
(-[WKContentView systemDidChangeSelectionForInteraction:selectionDidChange:]):
(-[WKContentView _showDictionary:]): Deleted.

Remove this altogether, since it&apos;s now just dead code with recent versions of UIKit.

(toWKSESelectionTouch): Deleted.
(-[WKContentView setAsyncSystemInputDelegate:]): Deleted.
(-[WKContentView handleAsyncKeyEvent:withCompletionHandler:]): Deleted.
(-[WKContentView replaceText:withText:options:withCompletionHandler:]): Deleted.
(-[WKContentView moveInDirection:byGranularity:]): Deleted.
(-[WKContentView extendInDirection:byGranularity:]): Deleted.
(-[WKContentView adjustSelection:completionHandler:]): Deleted.
(-[WKContentView shiftKeyStateChangedFrom:to:]): Deleted.
(-[WKContentView selectionWillChange:]): Deleted.
(-[WKContentView selectionDidChange:]): Deleted.
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.h:
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.mm:
(-[WKExtendedTextInputTraits setSelectionHandleColor:]):
(-[WKExtendedTextInputTraits selectionHandleColor]):
(-[WKExtendedTextInputTraits setSelectionBarColor:]):
(-[WKExtendedTextInputTraits selectionBarColor]):
(-[WKExtendedTextInputTraits setSelectionColorsToMatchTintColor:]):
* Source/WebKit/UIProcess/ios/WKSEDefinitions.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper initWithView:]):
(-[WKTextInteractionWrapper selectionChanged]):
(-[WKTextInteractionWrapper selectionChangedWithGestureAt:withGesture:withState:withFlags:]):
(-[WKTextInteractionWrapper selectionChangedWithTouchAt:withSelectionTouch:withFlags:]):
(-[WKTextInteractionWrapper lookup:withRange:fromRect:]):
(-[WKTextInteractionWrapper showShareSheetFor:fromRect:]):
(-[WKTextInteractionWrapper showTextServiceFor:fromRect:]):
(-[WKTextInteractionWrapper scheduleReplacementsForText:]):
(-[WKTextInteractionWrapper scheduleChineseTransliterationForText:]):
(-[WKTextInteractionWrapper translate:fromRect:]):
(-[WKTextInteractionWrapper showDictionaryFor:fromRect:]): Deleted.
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
* Tools/TestWebKitAPI/Tests/ios/AutocorrectionTestsIOS.mm:
(TEST):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView moveSelectionToStartOfParagraph]):
(-[WKWebView extendSelectionToStartOfParagraph]):
(-[WKWebView moveSelectionToEndOfParagraph]):
(-[WKWebView extendSelectionToEndOfParagraph]):
(-[WKWebView extendedTextInputTraits]):
(-[WKWebView autocorrectionContext]):
(-[WKWebView replaceText:withText:shouldUnderline:completion:]):
(wrap):
(unwrap):

Update `TestWKWebView` API testing helpers to use the new API shape when
`SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` is set.

(-[WKWebView synchronouslyAdjustSelectionWithDelta:]):
(-[WKWebView selectTextForContextMenuWithLocationInView:completion:]):
(-[WKWebView handleKeyEvent:completion:]):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::applyAutocorrection):

Canonical link: <a href="https://commits.webkit.org/273093@main">https://commits.webkit.org/273093@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1e1b2b0f260f24dccbd3df53fab6a73fd149188

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13031 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36211 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36911 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35306 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15418 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10161 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30036 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34734 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11036 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30542 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9632 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9743 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30579 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38201 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/29136 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31100 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30897 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35822 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/34146 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9861 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7727 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33745 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11648 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/40685 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7881 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10420 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8473 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10681 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -1193,7 +1189,7 @@ - (void)_updateRuntimeProtocolConformanceIfNeeded
             using MethodDescriptionList = objc_method_description[];
             auto methods = std::unique_ptr<MethodDescriptionList, WTF::SystemFree<MethodDescriptionList>> {
-                protocol_copyMethodDescriptionList(protocol, NO, YES, &methodCount)
+                protocol_copyMethodDescriptionList(protocol, YES, YES, &methodCount)
```
Did the protocol specification change? Or should these always have been required methods?

whsieh replied:
Ah, they're now required in the new text input protocol (whereas they weren't before). I went ahead and adjusted this to match that.

aprotyas reviewed (commented):
LGTM

whsieh committed:
[iOS] Update WKSETextInput/WKSETextInteraction adoption
https://bugs.webkit.org/show_bug.cgi?id=267555
rdar://121011826

Reviewed by Tim Horton.

Since 272018@main landed, there have been numerous minor adjustments to the async text input API in
ServiceExtensions; now that the API shape is finally stable, we adjust the WebKit preadoption to use
these new method and symbol names when `SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` is set, while
ensuring binary compatibility with the old UIKit async text input symbols.

See below for more details.

* Source/WebCore/platform/ios/WebEvent.mm:
(webEventType):
(isChangingKeyModifiers):
(-[WebEvent initWithKeyEntry:]):
(-[WebEvent originalKeyEntry]):
(-[WebEvent initWithKeyEvent:]): Deleted.
(-[WebEvent originalKeyEvent]): Deleted.

Rename `WebSEKeyEvent` to `WebSEKeyEntry`, and `WebSEKeyEventType` to `WebSEKeyPressState`.

* Source/WebCore/platform/ios/WebEventPrivate.h:
* Source/WebCore/platform/ios/WebSEDefinitions.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove `HAVE_UI_ASYNC_TEXT_INTERACTION_DELEGATE`, since it's been in SDKs where
`HAVE_UI_ASYNC_TEXT_INTERACTION` is set for months now.

* Source/WebKit/Shared/DocumentEditingContext.mm:

Import the relevant system headers.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Rename numerous text input delegate method implementations, to fit the new API shape.

(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):

Also address the fact that calls to `-hasSelectablePositionAtPoint:` have been replaced with calls
to `-textInteractionGesture:shouldBeginAtPoint:` with `WKSEGestureTypeForceTouch`, since the former
was only used in a few places for the purposes of handling force touch.

(-[WKContentView replaceForWebView:]):
(-[WKContentView selectionBarColor]):

Replace `-selectionBarColor` with `-selectionHandleColor`.

(-[WKContentView canPerformAction:withSender:]):

Replace `-moveInDirection:byGranularity:` with `-moveInStorageDirection:byGranularity:`.

(toGestureType):
(toWKSEGestureType):
(toSelectionTouch):
(toWKSESelectionTouchPhase):
(selectionChangedWithTouch):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(logTextInteraction):
(-[WKContentView _internalBeginSelectionChange]):
(-[WKContentView _internalEndSelectionChange]):
(-[WKContentView addTextAlternatives:]):
(-[WKContentView _updateTextInputTraits:]):
(shiftKeyState):
(-[WKContentView modifierFlagsDidChangeFrom:to:]):
(-[WKContentView _deferKeyEventToInputMethodEditing:]):
(-[WKContentView _interpretKeyEvent:isCharEvent:]):
(-[WKContentView _transpose]):
(-[WKContentView _updateInputContextAfterBlurringAndRefocusingElement]):
(-[WKContentView _selectionChanged]):
(-[WKContentView _provideSuggestionsToInputDelegate:]):
(-[WKContentView _provideUITextSuggestionsToInputDelegate:]):
(toWebDocumentRequestOptions):
(toWebRequest):
(-[WKContentView requestDocumentContext:completionHandler:]):
(-[WKContentView asWKSETextInput]):

The following method implementations represent the set of final API names, which wrap calls to
`UIAsyncTextInput` methods.

(-[WKContentView systemWillPresentEditMenuWithAnimator:]):
(-[WKContentView systemWillDismissEditMenuWithAnimator:]):
(-[WKContentView insertTextAlternatives:]):
(-[WKContentView insertText:textAlternatives:style:]):

Note: `-insertText:textAlternatives:style:` was renamed to `-insertTextAlternatives:` here, but we
implement both versions of this new delegate method for now to ensure binary compatibility with
both versions of ServiceExtensions.

(-[WKContentView isPointNearMarkedText:]):
(-[WKContentView replaceSelectedText:withText:]):
(-[WKContentView updateCurrentSelectionTo:fromGesture:inState:]):
(-[WKContentView setSelectionFromPoint:toPoint:gesture:state:]):
(-[WKContentView adjustSelectionBoundaryToPoint:touchPhase:baseIsStart:flags:]):
(-[WKContentView moveSelectionAtBoundary:inStorageDirection:completionHandler:]):
(-[WKContentView extendedTextInputTraits]):
(-[WKContentView selectTextForEditMenuWithLocationInView:completionHandler:]):
(-[WKContentView isSelectionAtDocumentStart]):
(-[WKContentView automaticallyPresentEditMenu]):
(-[WKContentView asyncSystemInputDelegate]):
(-[WKContentView setAsyncInputDelegate:setAsyncSystemInputDelegate:]):
(-[WKContentView handleKeyEntry:withCompletionHandler:handleAsyncKeyEvent:withCompletionHandler:]):
(-[WKContentView replaceText:withText:options:completionHandler:replaceText:withText:options:withCompletionHandler:]):
(-[WKContentView moveInStorageDirection:byGranularity:moveInDirection:byGranularity:]):
(-[WKContentView extendInStorageDirection:byGranularity:extendInDirection:byGranularity:]):
(-[WKContentView adjustSelectionByRange:completionHandler:adjustSelection:completionHandler:]):
(-[WKContentView transposeCharacters]):
(-[WKContentView shiftKeyStateChangedFromState:toState:shiftKeyStateChangedFrom:to:]):
(-[WKContentView systemWillChangeSelectionForInteraction:selectionWillChange:]):
(-[WKContentView systemDidChangeSelectionForInteraction:selectionDidChange:]):
(-[WKContentView _showDictionary:]): Deleted.

Remove this altogether, since it's now just dead code with recent versions of UIKit.

(toWKSESelectionTouch): Deleted.
(-[WKContentView setAsyncSystemInputDelegate:]): Deleted.
(-[WKContentView handleAsyncKeyEvent:withCompletionHandler:]): Deleted.
(-[WKContentView replaceText:withText:options:withCompletionHandler:]): Deleted.
(-[WKContentView moveInDirection:byGranularity:]): Deleted.
(-[WKContentView extendInDirection:byGranularity:]): Deleted.
(-[WKContentView adjustSelection:completionHandler:]): Deleted.
(-[WKContentView shiftKeyStateChangedFrom:to:]): Deleted.
(-[WKContentView selectionWillChange:]): Deleted.
(-[WKContentView selectionDidChange:]): Deleted.
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.h:
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.mm:
(-[WKExtendedTextInputTraits setSelectionHandleColor:]):
(-[WKExtendedTextInputTraits selectionHandleColor]):
(-[WKExtendedTextInputTraits setSelectionBarColor:]):
(-[WKExtendedTextInputTraits selectionBarColor]):
(-[WKExtendedTextInputTraits setSelectionColorsToMatchTintColor:]):
* Source/WebKit/UIProcess/ios/WKSEDefinitions.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper initWithView:]):
(-[WKTextInteractionWrapper selectionChanged]):
(-[WKTextInteractionWrapper selectionChangedWithGestureAt:withGesture:withState:withFlags:]):
(-[WKTextInteractionWrapper selectionChangedWithTouchAt:withSelectionTouch:withFlags:]):
(-[WKTextInteractionWrapper lookup:withRange:fromRect:]):
(-[WKTextInteractionWrapper showShareSheetFor:fromRect:]):
(-[WKTextInteractionWrapper showTextServiceFor:fromRect:]):
(-[WKTextInteractionWrapper scheduleReplacementsForText:]):
(-[WKTextInteractionWrapper scheduleChineseTransliterationForText:]):
(-[WKTextInteractionWrapper translate:fromRect:]):
(-[WKTextInteractionWrapper showDictionaryFor:fromRect:]): Deleted.
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
* Tools/TestWebKitAPI/Tests/ios/AutocorrectionTestsIOS.mm:
(TEST):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView moveSelectionToStartOfParagraph]):
(-[WKWebView extendSelectionToStartOfParagraph]):
(-[WKWebView moveSelectionToEndOfParagraph]):
(-[WKWebView extendSelectionToEndOfParagraph]):
(-[WKWebView extendedTextInputTraits]):
(-[WKWebView autocorrectionContext]):
(-[WKWebView replaceText:withText:shouldUnderline:completion:]):
(wrap):
(unwrap):

Update `TestWKWebView` API testing helpers to use the new API shape when
`SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` is set.

(-[WKWebView synchronouslyAdjustSelectionWithDelta:]):
(-[WKWebView selectTextForContextMenuWithLocationInView:completion:]):
(-[WKWebView handleKeyEvent:completion:]):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::applyAutocorrection):

Canonical link: https://commits.webkit.org/273093@main
